### PR TITLE
Avoid unnecessary copying of non-cheaply-copied types

### PR DESCRIPTION
**Pull request by practicalswift**
* Avoid unnecessary copying of non-cheaply-copied types. Use pass by reference to const consistently where appropriate.
* Don't use pass by reference to const for cheaply-copied types (bool, char, etc.).

### CODE REVIEW

**Comment by promag** [comment]
PR title too long. Please keep it short and clear.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -1683,7 +1683,7 @@ CAmount CWalletTx::GetCredit(const isminefilter& filter) const
     return credit;
 }
 
-CAmount CWalletTx::GetImmatureCredit(bool fUseCache) const
+CAmount CWalletTx::GetImmatureCredit(const bool fUseCache) const
```
Comment (promag) [inline src/wallet/wallet.cpp]: Why these changes?

**Review by promag — COMMENTED** [review]
What are the non-cheaply-copied types? Have you measured the difference?

Also, this is incomplete, for instance, for type `uint256`:
```cpp
bool bumpFee(uint256 hash);
```
could be
```cpp
bool bumpFee(const uint256& hash);
```

**Comment by sipa** [comment]
@promag Generally, anything larger than 1 or 2 pointers is better passed by reference.

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (practicalswift) [inline src/wallet/wallet.cpp]: To get signature consistency with `GetImmatureWatchOnlyCredit`. See the changes made to the same file.

**Comment by practicalswift** [comment]
@promag See @sipa:s answer for the common definition of (non-)cheaply-copied types.

Regarding completeness: I only targeted `const` arguments.

File: src/wallet/feebumper.cpp [hunk 2]
```diff
@@ -66,7 +66,7 @@ bool CFeeBumper::preconditionChecks(const CWallet *pWallet, const CWalletTx& wtx
     return true;
 }
 
-CFeeBumper::CFeeBumper(const CWallet *pWallet, const uint256 txidIn, int newConfirmTarget, bool ignoreGlobalPayTxFee, CAmount totalFee, bool newTxReplaceable)
+CFeeBumper::CFeeBumper(const CWallet *pWallet, const uint256& txidIn, int newConfirmTarget, bool ignoreGlobalPayTxFee, CAmount totalFee, bool newTxReplaceable)
```
Comment (TheBlueMatt) [inline src/wallet/feebumper.cpp]: This is broken - txidIn is std::move()d two lines down.

**Review by TheBlueMatt — COMMENTED** [review]
These kinds of changes scare me, see the bug introduced in CFeeBumper. Some of them are obvious, some of them could subtley introduce bugs (did you look deeply at the CTxMempool changes?).

**Comment by MIGUELWAXX** [comment]
cheers

**Comment by MIGUELWAXX** [comment]
cheers

**Comment by practicalswift** [comment]
@TheBlueMatt Good point! I missed the `std::move(…)`. Closing this PR - I'll have to evaluate it deeper and contemplate if the performance reward is worth it. Thanks for reviewing!

File: src/wallet/feebumper.cpp (hunk 2 above)
Comment (sipa) [inline src/wallet/feebumper.cpp]: It certainly seems misplaced, but it's not actually wrong, as std::move in this case just casts to `const uint256&&`, which won't be moved from (also, uint256 doesn't have a move constructor that destroys the argument).

Comment (dcousens) [inline src/wallet/feebumper.cpp]: `const uint256&` will not be moved,  `std::move` is simply a cast,  it will do nothing here.

Comment (ArielLorenzo-Luaces) [inline src/wallet/feebumper.cpp]: ```txidIn``` should not be const nor pass-by-reference.
If ```CFeeBumper``` is going to cannibalize ```txidIn``` two lines down it should be copied in the case it was an lvalue. In the case that it was an rvalue, it will not be copied anyway.
```const uint256&``` only causes inefficiencies (due to unnecessary copying) when the caller sends ```txidIn``` as an rvalue

Comment (TheBlueMatt) [inline src/wallet/feebumper.cpp]: @sipa sorry, broken is a bad word there, I meant "unsafe code", ie we want to be careful of having code like that.
